retworkx-core/src/planar/lr_planar.rs

```diff
@@ -0,0 +1,699 @@
+    fn lowest<G>(&self, lr_state: &LRState<G>) -> usize
+    where
+        G: GraphBase<NodeId = T>,
+    {
+        match (self.left.low(), self.right.low()) {
+            (Some(l_low), Some(r_low)) => lr_state.lowpt[l_low].min(lr_state.lowpt[r_low]),
+            (Some(l_low), None) => lr_state.lowpt[l_low],
+            (None, Some(r_low)) => lr_state.lowpt[r_low],
+            (None, None) => std::usize::MAX,
+        }
+    }
+}
+
+enum Sign {
+    Plus,
+    Minus,
+}
+
+/// Similar to ``DfsEvent`` plus an extra event ``FinishEdge``
+/// that indicates that we have finished processing an edge.
+enum LRTestDfsEvent<N> {
+    Finish(N),
+    TreeEdge(N, N),
+    BackEdge(N, N),
+    FinishEdge(N, N),
+}
+
+// An error: graph is *not* planar.
+struct NonPlanar {}
+
+struct LRState<G: GraphBase>
+where
+    G::NodeId: Hash + Eq,
+{
+    graph: G,
+    /// roots of the DFS forest.
+    roots: Vec<G::NodeId>,
+    /// distnace from root.
+    height: HashMap<G::NodeId, usize>,
+    /// parent edge.
+    eparent: HashMap<G::NodeId, Edge<G>>,
+    /// height of lowest return point.
+    lowpt: HashMap<Edge<G>, usize>,
+    /// height of next-to-lowest return point. Only used to check if an edge is chordal.
+    lowpt_2: HashMap<Edge<G>, usize>,
+    /// next back edge in traversal with lowest return point.
+    lowpt_edge: HashMap<Edge<G>, Edge<G>>,
+    /// proxy for nesting order ≺ given by twice lowpt (plus 1 if chordal).
+    nesting_depth: HashMap<Edge<G>, usize>,
+    /// stack for conflict pairs.
+    stack: Vec<ConflictPair<Edge<G>>>,
+    /// marks the top conflict pair when an edge was pushed in the stack.
+    stack_emarker: HashMap<Edge<G>, ConflictPair<Edge<G>>>,
+    /// edge relative to which side is defined.
+    eref: HashMap<Edge<G>, Edge<G>>,
+    /// side of edge, or modifier for side of reference edge.
+    side: HashMap<Edge<G>, Sign>,
+}
+
+impl<G> LRState<G>
+where
+    G: GraphBase + NodeCount + EdgeCount + IntoEdges + Visitable,
+    G::NodeId: Hash + Eq,
+{
+    fn new(graph: G) -> Self {
+        let num_nodes = graph.node_count();
+        let num_edges = graph.edge_count();
+
+        LRState {
+            graph,
+            roots: Vec::new(),
+            height: HashMap::with_capacity(num_nodes),
+            eparent: HashMap::with_capacity(num_edges),
+            lowpt: HashMap::with_capacity(num_edges),
+            lowpt_2: HashMap::with_capacity(num_edges),
+            lowpt_edge: HashMap::with_capacity(num_edges),
+            nesting_depth: HashMap::with_capacity(num_edges),
+            stack: Vec::new(),
+            stack_emarker: HashMap::with_capacity(num_edges),
+            eref: HashMap::with_capacity(num_edges),
+            side: graph
+                .edge_references()
+                .map(|e| ((e.source(), e.target()), Sign::Plus))
+                .collect(),
+        }
+    }
+
+    fn lr_orientation_visitor(&mut self, event: DfsEvent<G::NodeId, &G::EdgeWeight>) {
```

Comment:
It appears that you are following a different structure from `networkx`, using an event handler rather than a more procedural approach. Is there a reason for that?

Reply:
It's mostly a personal preference. It's done since this PR makes use of [`depth_first_search`](https://github.com/Qiskit/retworkx/blob/main/retworkx-core/src/traversal/dfs_visit.rs#L159) that pre-existed and follows this approach.
In general I think it improves:
- Code readability. You develop/review the graph traversal and the custom logic inserted through the event handlers separately.
- Code reusability. We can re-use `lr_visit_ordered_dfs_tree` and implement a new `fn lr_embedding_visitor(&mut self, event: DfsEvent<G::NodeId, &G::EdgeWeight>)` event handler that will help in creating the combinatorial embedding.